src/tools/metrics/summary.rs

```diff
@@ -0,0 +1,92 @@
+use core::fmt;
+use std::{
+    collections::HashMap,
+    net::SocketAddr,
+    time::{Instant, Duration},
+    fs::File, 
+    io::Write,
+    cmp,
+};
+
+use crate::tools::crawler::KnownNode;
+
+#[allow(dead_code)]
+pub struct NetworkSummary {
+    num_known_nodes: usize,
+    num_good_nodes: usize,
+    protocol_versions: HashMap<u32, usize>,
+    user_agents: HashMap<String, usize>,
+    crawler_runtime: Duration,
+}
+
+impl NetworkSummary {
+    /// Constructs a new NetworkSummary from given nodes.
+    pub fn new(nodes: HashMap<SocketAddr, KnownNode>, crawler_start_time: Instant) -> NetworkSummary {
+        let num_known_nodes = nodes.len();
+
+        let good_nodes: HashMap<SocketAddr, KnownNode> = nodes.clone()
```

Comment:
```suggestion
        let good_nodes: HashMap<_, _> = nodes.clone()
```
this can likely be elided